### PR TITLE
Merge nvme cli monolithic

### PR DESCRIPTION
**Pull request by igaw**
This is a first rough version attempt to merge the missing patches from the nvme-cli-monolithic branch. Please see
also my notes in my PR in nvme-cli https://github.com/linux-nvme/nvme-cli/pull/1223

Ah yes, most stuff is only compile tested...

### CODE REVIEW

**Comment by igaw** [comment]
@keithbusch  and @hreinecke please double check if the naming convention I used is okay. I had to adapt from the
original patch. So it's likely off in places.

**Comment by keithbusch** [comment]
Thanks for doing this merge! Looks okay at first glance. I am traveling this week, though, I'll have time to test things out on Monday.

**Review by hreinecke — APPROVED** [review]
Looks good.

**Review by hreinecke — COMMENTED** [review]
Hmm. This will change the ABI, and will cause nvme-cli to not build against this version unless patched, too.
Are we okay with that? Or do we need to provide compability wrapper?

**Comment by igaw** [comment]
There was no official libnvme release yet, no? I think we can still argue that his doesn't need any backwards compatibility  (yet)...

File: src/libnvme.map [hunk 1]
```diff
@@ -114,6 +114,7 @@
 		nvme_get_log_endurance_group;
 		nvme_get_log_endurance_grp_evt;
 		nvme_get_log_error;
+		nvme_get_log_fid_sup_efcs;
 		nvme_get_log_fw_slot;
```
Comment (hreinecke) [inline src/libnvme.map:118]: Please make that 'get_log_fid_sup_effects'. 'efcs' is making no sense, and is not something one could figure out as an abbreviation.

File: src/nvme/ioctl.h [hunk 2]
```diff
@@ -1004,6 +1004,18 @@ int nvme_get_log_lba_status(int fd, bool rae, __u64 offset, __u32 len,
 int nvme_get_log_endurance_grp_evt(int fd, bool rae, __u32 offset, __u32 len,
 				   void *log);
 
+/**
+ * nvme_get_log_fid_sup_efcs() -
+ * @fd:		File descriptor of nvme device
```
Comment (hreinecke) [inline src/nvme/ioctl.h:1032]: Same here.

**Review by hreinecke — CHANGES_REQUESTED** [review]

File: src/nvme/ioctl.c [hunk 3]
```diff
@@ -1927,8 +1927,8 @@ int nvme_resv_report(int fd, __u32 nsid, bool eds, __u32 len,
 }
 
 int nvme_zns_mgmt_send(int fd, __u32 nsid, __u64 slba, bool select_all,
-		       enum nvme_zns_send_action zsa, __u32 data_len,
-		       void *data)
+		       __u32 timeout, enum nvme_zns_send_action zsa,
+		       __u32 data_len, void *data)
```
Comment (hreinecke) [inline src/nvme/ioctl.c:1963]: Odd. Why do have zns command a timeout, and other commands haven't?
I'm pretty sure we can set a timeout on any command, so what makes these special?

File: src/nvme/types.h [hunk 4]
```diff
@@ -2785,7 +2787,9 @@ struct nvme_telemetry_log {
 	__le16	dalb1;
 	__le16	dalb2;
 	__le16	dalb3;
-	__u8	rsvd14[368];
+	__u8    rsvd14[2];
+	__le32  dalb4;
+	__u8    rsvd20[362];
 	__u8	ctrlavail;
```
Comment (hreinecke) [inline src/nvme/types.h:2809]: All good, but where's the plugin?
Shouldn't we modify the commit message to refer to the changes we actually did?

**Review by hreinecke — CHANGES_REQUESTED** [review]

File: src/nvme/ioctl.c (hunk 3 above)
Comment (igaw) [reply src/nvme/ioctl.c:1963]: I don't know, this is from  fa97500b9a72 ("Add Timeout Parameter for Zone Management Commands")
https://github.com/linux-nvme/nvme-cli/commit/fa97500b9a72

**Comment by hreinecke** [comment]
Hmm. Okay, just wanted to point out. Which means it's just the one commit message which needs to be modified. Rest is okay for merging.

File: src/libnvme.map (hunk 1 above)
Comment (igaw) [reply src/libnvme.map:118]: I was also unhappy with it, especially after I realized that I spelled it NVME_LOG_LID_FID_SUPPORT_EFFECTS for the define.

File: src/nvme/ioctl.h (hunk 2 above)
Comment (igaw) [reply src/nvme/ioctl.h:1032]: oki

File: src/nvme/types.h (hunk 4 above)
Comment (igaw) [reply src/nvme/types.h:2809]: The plugin part is in the nvme-cli PR (https://github.com/linux-nvme/nvme-cli/pull/1223):

https://github.com/linux-nvme/nvme-cli/pull/1223/commits/d361c3781778e658fb1f67109847ebf53111ecd9

**Comment by igaw** [comment]
New version, rebased on current master and changed following names:

```
nvme_get_log_sup_pages    -> nvme_get_log_supported_log_pages                      
nvme_get_log_fid_sup_efcs -> nvme_get_log_fid_supported_effects                    
nvme_cap_mgmt             -> nvme_capacity_mgmt       
```
This should make things more consistent with the rest of the code base

File: src/nvme/ioctl.c (hunk 3 above)
Comment (igaw) [reply src/nvme/ioctl.c:1963]: I suppose we should add the timeout to all commands then. I'll created an issue to track this, see #97

File: src/nvme/types.h (hunk 4 above)
Comment (igaw) [reply src/nvme/types.h:2809]: Ah got it, I changed the commit message title. Indeed a bit confusing.

**Comment by igaw** [comment]
If no one is vetoing this down now, I am going to apply it soon (after processing the pending PRs).

**Comment by igaw** [comment]
FYI, just a rebase to recent HEAD to do a final check before merge

**Pull request merged**